BilateralInteractionConstraint issue using sofa v21.06

### Problem

**Description**
BilateralInteractionConstraint doesn't work for me anymore when I upgraded to sofa v21.06.

**Steps to reproduce**
v21.06 builds well on my pc.
In the demo/caduceus.scn I set the Mouse left button to  “attach object using Bilateral constraint” but I can not grasp the snake using the mouse. (see screenshot attached)
![bilateral_notworking_snake](https://user-images.githubusercontent.com/14253725/142281753-285ebe01-e481-4f43-87d7-51e181d3368e.png)

Another example is in my scene i attached the gallbladder model to the fatty tissue behind it using Bilateral constraint, i checked the rendering interactions, which visualize the Bilateral constraint contact pairs using Pink dots. Based on the screenshot, the contact points are correctly visualized but they can still move apart as if no constraints at all. (see screenshot attached)

![bilateral_notworking](https://user-images.githubusercontent.com/14253725/142281781-13222f7d-7843-4a38-b7c2-9e93a40dc6a9.png)

**Expected behavior**
In our plugins we have used the following to establish the bilaterial constraints.
  _sofa::component::constraintset::BilateralInteractionConstraint<DataTypes>::SPtr 
  toolState.m_constraints->addContact(c.normal, c.point[0], c.point[1], distance, index1, index2, c.point[0], c.point[1], 1, c.id);_

The same codes and scene files worked well in sofa **v19.12**, where the bilateral constraints worked as constraining the distances between the contact pairs.

---------------------------------------------

### Environment

**Context**

- System: Windows 10
- Version of SOFA:  branch v21.06.00
- State: Build directory

**Command called**

```txt

< COPY-PASTE YOUR COMMAND HERE >

```

**Env vars**

```bash
python -c "exec( \"import os, sys\nprint('#################')\nprint('--- sys.version ---')\nprint(sys.version)\nprint('--- PATH ---')\ntry:\n  print(os.environ['PATH'])\nexcept Exception:\n  pass\nprint('--- SOFA_ROOT ---')\ntry:\n  print(os.environ['SOFA_ROOT'])\nexcept Exception:\n  pass\nprint('--- PYTHONPATH ---')\ntry:\n  print(os.environ['PYTHONPATH'])\nexcept Exception:\n  pass\nprint('--- sys.path ---')\ntry:\n   print(str(sys.path))\nexcept Exception:\n   pass\nprint('#################')\" )"
```

```txt

#################
--- sys.version ---
2.7.18 (v2.7.18:8d21aa21f2, Apr 20 2020, 13:25:05) [MSC v.1500 64 bit (AMD64)]
--- PATH ---
C:\Program Files\Eclipse Foundation\jdk-8.0.302.8-hotspot\bin;C:\Python27;C:\Users\Ruiliang\AppData\Local\Programs\Python\Python39\Scripts;C:\Users\Ruiliang\AppData\Local\Programs\Python\Python39;C:\SIMULIA\Commands;C:\Program Files (x86)\Razer\ChromaBroadcast\bin;C:\Program Files\Razer\ChromaBroadcast\bin;C:\Program Files (x86)\Razer Chroma SDK\bin;C:\Program Files\Razer Chroma SDK\bin;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Windows\System32\OpenSSH\;C:\Program Files\Crucial\Crucial Storage Executive;C:\Program Files\NVIDIA Corporation\NVIDIA NvDLISR;C:\Windows\system32\config\systemprofile\AppData\Local\Microsoft\WindowsApps;C:\Users\Ruiliang\AppData\Local\Microsoft\WindowsApps;C:\Users\Ruiliang\AppData\Local\Programs\Microsoft VS Code\bin;C:\Program Files\CMake\bin;C:\Qt\5.15.2\msvc2019_64\bin;C:\Qt\5.15.2\msvc2019_64\lib;C:\local\boost_1_75_0\lib64-msvc-14.2;C:\local\boost_1_75_0;C:\Program Files\Git\cmd;;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\WINDOWS\System32\WindowsPowerShell\v1.0\;C:\WINDOWS\System32\OpenSSH\;C:\Program Files\dotnet\;C:\Users\Ruiliang\AppData\Local\Programs\Python\Python39\Scripts\;C:\Users\Ruiliang\AppData\Local\Programs\Python\Python39\;C:\Users\Ruiliang\AppData\Local\Microsoft\WindowsApps;C:\Users\Ruiliang\AppData\Local\Programs\Microsoft VS Code\bin;C:\Users\Ruiliang\.dotnet\tools
--- SOFA_ROOT ---
--- PYTHONPATH ---
--- sys.path ---
['', 'C:\\WINDOWS\\SYSTEM32\\python27.zip', 'C:\\Python27\\DLLs', 'C:\\Python27\\lib', 'C:\\Python27\\lib\\plat-win', 'C:\\Python27\\lib\\lib-tk', 'C:\\Python27', 'C:\\Python27\\lib\\site-packages']
#################

```

---------------------------------------------

### Logs

**Full output**

```txt

No Error outputs or suspicious outputs has been found.

```

**Content of build_dir/CMakeCache.txt**

< DRAG AND DROP YOUR CMAKECACHE.TXT HERE >

---------------------------------------------

Thank you for your report.

@rgao15 Thanks a lot for the feedback!

I tried to reproduce your issue. On my end, I do have an error message:
```
[ERROR]   [MouseOperation] ConstraintAttachBody performer cannot be created with the picked model.
```
Let us know if you have the same error message. I am also on Windows.

Anyway, I tried to understand what's happening. Why ConstraintAttachBody cannot be created?
It is supposed to be added in a factory here : https://github.com/sofa-framework/sofa/blob/323e20e4168e1ab5b67352f4621d4f1b3fa39be9/modules/SofaConstraint/src/SofaConstraint/ConstraintAttachBodyPerformer.cpp#L38
Registration in the factory works BUT not in the right factory.
The factory in `modules/SofaGuiCommon/src/sofa/gui/MouseOperations.cpp` and in `modules/SofaConstraint/src/SofaConstraint/ConstraintAttachBodyPerformer.cpp` are different. The factory is a singleton but there are 2 different instances. I confirmed it by printing the factory address. I suspect it's because they are used in different libraries.

The question is: how to solve it? And is it the same story for other factories? @fredroy @damienmarchal I would need your help

--- a/modules/SofaConstraint/SofaConstraint_test/CMakeLists.txt
+++ b/modules/SofaConstraint/SofaConstraint_test/CMakeLists.txt
@@ -7,10 +7,11 @@ sofa_find_package(SofaConstraint REQUIRED)
 set(SOURCE_FILES ../../empty.cpp)
 
 list(APPEND SOURCE_FILES
-    #LocalMinDistance_test.cpp
-    GenericConstraintSolver_test.cpp
     BilateralInteractionConstraint_test.cpp
-    UncoupledConstraintCorrection_test.cpp)
+    ConstraintAttachBodyPerformer_test.cpp
+    GenericConstraintSolver_test.cpp
+    UncoupledConstraintCorrection_test.cpp
+)
 
 add_definitions("-DSOFATEST_SCENES_DIR=\"${CMAKE_CURRENT_SOURCE_DIR}/scenes_test\"")
 add_executable(${PROJECT_NAME} ${SOURCE_FILES})

--- a/modules/SofaConstraint/SofaConstraint_test/ConstraintAttachBodyPerformer_test.cpp
+++ b/modules/SofaConstraint/SofaConstraint_test/ConstraintAttachBodyPerformer_test.cpp
@@ -1,0 +1,35 @@
+/******************************************************************************
+*                 SOFA, Simulation Open-Framework Architecture                *
+*                    (c) 2006 INRIA, USTL, UJF, CNRS, MGH                     *
+*                                                                             *
+* This program is free software; you can redistribute it and/or modify it     *
+* under the terms of the GNU Lesser General Public License as published by    *
+* the Free Software Foundation; either version 2.1 of the License, or (at     *
+* your option) any later version.                                             *
+*                                                                             *
+* This program is distributed in the hope that it will be useful, but WITHOUT *
+* ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or       *
+* FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License *
+* for more details.                                                           *
+*                                                                             *
+* You should have received a copy of the GNU Lesser General Public License    *
+* along with this program. If not, see <http://www.gnu.org/licenses/>.        *
+*******************************************************************************
+* Authors: The SOFA Team and external contributors (see Authors.txt)          *
+*                                                                             *
+* Contact information: contact@sofa-framework.org                             *
+******************************************************************************/
+#include <SofaConstraint/initSofaConstraint.h>
+#include <gtest/gtest.h>
+#include <SofaUserInteraction/InteractionPerformer.h>
+
+namespace sofa
+{
+
+TEST(ConstraintAttachBodyPerformer, Factory)
+{
+    component::initSofaConstraint();
+    EXPECT_TRUE(component::collision::InteractionPerformer::InteractionPerformerFactory::HasKey("ConstraintAttachBody"));
+}
+
+}

--- a/modules/SofaUserInteraction/src/SofaUserInteraction/InteractionPerformer.h
+++ b/modules/SofaUserInteraction/src/SofaUserInteraction/InteractionPerformer.h
@@ -78,3 +78,8 @@ public:
 };
 
 } //namespace sofa::component::collision
+
+namespace sofa::helper
+{
+extern template class SOFA_SOFAUSERINTERACTION_API Factory<std::string, component::collision::InteractionPerformer, component::collision::BaseMouseInteractor*>;
+} //namespace sofa::helper

